### DISCUSSION TITLE
formik 참고 자료

### DISCUSSION THREAD

**Discussion by Lim-JiSeon** [Team-PalPalHae-Dealight/Team-PalPalHae-Dealight-FE#93] [Show and tell]
# formik 이란?
form 라이브러리로, [Formik 공식문서](https://formik.org/docs/overview) 에서는 Formik이 가지는 장점을 아래와 같이 설명합니다.

- Getting values in and out of form state: 값 상태 관리
- Validation and error messages: 유효성 검사
- Handling form submission: 제출 방식 제어
- 
Formik은 단순히 input component 관리를 넘어서서 **데이터의 흐름을 제어**할 수 있는데 효과적이라고 표현합니다. 
Formik을 사용하면 마치 **Context API를 사용하는 것처럼** form이 필요한 컴포넌트를 <Formik> 태그로 감싸주어 props drilling 없이 '**컴포넌트 내에서 전역적으로 상태 관리**'를 할 수 있습니다. 
또한 form에서 늘 빠질 수 없는 **유효성 검사**까지 같이 해줄 수 있다는 장점이 있습니다.

# formik 과 react hook form 차이점
- 코드 라인 및 글자 수
코드 라인은 formik이 더 적으며, 코드 글자 수는 react hook form 이 더 적습니다.(전체적으로 **formik**이 코드를 더 간결하게, 직관적으로 짤 수 있음)
- 생태계 비교
![image](https://github.com/Team-PalPalHae-Dealight/Team-PalPalHae-Dealight-FE/assets/83554018/6e26a193-b9c8-4c1a-852c-5fe8b5a91dff)
react hook form이 가장 우수하며, 그 뒤를 formik이 따르고 있습니다. 
2022년도까지는 formik이 더 선호도가 높았다면 2023년도에 이르러 **react hook form**이 더 높아졌습니다.
- 번들 크기 비교
![image](https://github.com/Team-PalPalHae-Dealight/Team-PalPalHae-Dealight-FE/assets/83554018/05fd3dfe-2eb4-4185-a265-1ff2a0f20e15)
**react-hook-form**이 가장 작은 것을 확인할 수 있으며, 그 다음으로  formik이 잇따릅니다.
- 종속성 비교
![image](https://github.com/Team-PalPalHae-Dealight/Team-PalPalHae-Dealight-FE/assets/83554018/e9c49d0f-8908-47df-9e81-324094805a76)
**react hook form**은 종속성이 없으며, formik은 약 7~8개 정도의 종속성을 지닙니다.
- 실행 속도
performance 측면이나 network 측면에서 모두 **react hook form**이 우수하나 이는 상대적으로 비교했을 때 가장 빠른 것일 뿐 그 외의 formik이나 rc-field-form과 같은 다른 form 라이브러리도 우수합니다. 
- 리렌더링 비교
압도적으로 **react hook form** 이 리렌더링 횟수가 적은 편입니다.
![image](https://github.com/Team-PalPalHae-Dealight/Team-PalPalHae-Dealight-FE/assets/83554018/10a0b395-852f-471b-9829-12a36a95eaac)
### 요약
생태계 측면에서, 생태계의 크기에서 react-hook-form이냐, formik냐 2개의 선택지로 좁힐 수 있습니다.
사용성 측면에서, formik에서 제공하는 컴포넌트를 사용할 경우 코드 양을 줄일 수 있습니다.
성능 측면에서, react-hook-form은 번들 크기가 작아 첫 렌더링 속도가 빠르고, 리렌더링 횟수가 월등히 적기 때문에 성능 측면으로 본다면 react-hook-form이 더 좋은 선택입니다.
유연성 측면에서, react-hook-form은 Uncontrolled 방식으로 성능에 이점이 있지만 폼 값을 컨트롤하기 어려운 측면이 있습니다. 이 단점을 보완하기 위해 watch라는 기능이 존재하지만, formik는 Controlled 방식이기 때문에 폼 값을 좀 더 유연성 있게 다룰 수 있어 유연성 측면에서는 formik가 더 좋은 선택으로 보입니다.(watch를 사용해 react hook form을 controlled 처럼 이용할 수 있으나 controlled를 지향하는 react에서 추천하는 방식은 아닙니다.)

**Formik은 리액트가 권장하는 Controlled Component를 만드는 것에 힘썼다면, React-Hook-Form은 퍼포먼스 최적화에 힘을 실었다고 할 수 있습니다.**

# formik 장단점 정리
- 장점
폼 값을 **유연성**있게 다룰 수 있으며, state 방식과  **context 방식이기 때문에 직관적이고 쉽게 데이터 흐름을 제어**할 수 있습니다. 추가적으로 controlled component 방식이기 때문에 context api처럼 데이터를 다룰 수 있어 하나의 컴포넌트로 여러 곳에서 입력받은 데이터들을 부모 컴포넌트에서(<Formik>) **전역으로 관리**할 수 있습니다.
**코드를 간결하게** 구현할 수 있으며, **validation 처리**하는 데 있어 매우 편리한 기능을 제공합니다.(error, validate, <ErrorMessage />  등)
- 단점
react hook form 보다는 성능 측면에서 떨어집니다.(그 외에는 큰 차이가 없으나 리렌더링에서는 차이가 확실함)
부분적인 watch 처리가 불가능합니다.(부분적으로 렌더링할 대상 결정 불가능)

### 딜라잇에서 formik 사용한 이유
- 다양한 방식으로 데이터를 입력받기 때문에 react 친화적이며 유연성이 높은 formik을 이용했습니다.(text, password, checkbox, select  등등)
- 입력 폼의 종류가 다양하기 때문에 코드의 간결성이 필요했으며, 폼에서 입력받는 값들이 많기 때문에 데이터 흐름을 제어하는데 편리한 formik을 사용했습니다.(회원가입, 상품 등록, 업체 등록, 마이페이지 등등)
- 다양한 validation을 확인해야하기 때문에 이러한 측면에서 다양한 기능들을 제공하고 편의성을 제공하는 formik을 사용했습니다.(예: 사업자 등록번호 같은 경우 총 10자의 숫자로 입력받아야 하며, 필수 입력, 지역코드, 개인/사업 코드, 지역코드 등 양식과 규격이 다양하게 정해져 있음)
- (이 부분은 추후 여유가 될 때 리팩토링하면서 formik의 장점을 확실하게 살릴려고 합니다. - 현재 미구현) controlled 방식이기 때문에 context 처럼 데이터를 전역으로 사용할 수 있어 formik을 사용했습니다.(업체 등록하기 같은 경우 여러 곳에서 데이터를 각기 입력받은 뒤 서버에 한 번에 입력받은 데이터들을 모아서 보내야 합니다. 이때 전역 상태로 데이터를 관리할 수 있는 formik이 유용하게 쓰일 수 있었습니다.)

# formik 사용 방법 예제 정리
### formik 사용 방법
먼저, <Formik> 컴포넌트가 가장 외곽이 됩니다. 이는, initialValues, onSubmit 2가지 Props를 기본적으로 요구받습니다.

<< Formik Props >>
굵은 글씨만 참고해주셔도 괜찮습니다.

- **initialValues** : 폼 내에서 쓸 데이터의 기본값. 객체를 받으며, 각 key는 내부 input의 name과 대응된다.
- **onSubmit** : 폼에서 Submit 이벤트가 발생했을 때 실행할 핸들러 함수를 등록한다.
인자는 values(입력값 객체), actions(폼의 에러상태, 제출중인 상태(submitting), 폼 리셋 등의 추가조작) 2가지를 받는다.
- 이외에도, validate(유효성 검사 함수), validatoinSchema(유효성 검사 객체) 등 인자를 받는다.(yup을 사용해 schema로 처리하면 더 간단하게 처리 가능)

<< Render Props >>
- **values** : <input> 등의 각 name과 이에 해당하는 value들의 값을 저장하는 객체. 각 name을 key로 값을 조회 가능하며, 위의initialValutes로 초기화.
- **errors** : validate 함수로 반환한 에러 정보를 저장하는 객체. 각 name을 key로 에러 메세지 조회 가능 (메세지가 있으면 에러)
- handleSubmit : 위의 onSubmit에 등록한 핸들러 함수. <form>의 onSubmit 이벤트 핸들러로 등록
- handleChange : 내부 엘리먼트의 값 변경을 연동하는 핸들러 함수. <input> 등의 onChange 이벤트 핸들러로 등록
- handleBlur : 내부 엘리먼트의 focusOut 동작과 연동하는 핸들러 함수(에러 핸들링). <input> 등의 onBlur 이벤트 핸들러로 등록
- touched : 사용자가 내부 엘리먼트와 인터렉션이 있었는지 여부를 저장하는 객체. 각 name을 key로 확인
- isSubmitting : 위의 onSubmit 핸들러로 설정한 setSubmitting의 true/false 여부. 비동기 통신중임을 표현하기 위해 활용
- 이외에도, dirty(initialValues와 깊은복사 일치 여부), isValid(유효성 여부), 및 value, error, 각 필드에 대한 제어 메서드 등이 존재

### formik 사용 예제
```ts
const initialValues = { // values의 초기값을 설정해주세요.(폼을 통해 입력받을 값들 초기값)
    storeNumber: '',
    name: 0,
    name: false, // 이런 식으로 name 자리에 변수명을 작성해주시고, 초기값을 넣어줏히면 됩니다.
  };
  const schema = object().shape({ // yup을 이용한 validation 
      storeNumber: string() // name 자리에 validation할 변수명을 넣어주고, 그 변수의 타입을 지정해주세요. 예시로 string(), number(), object() 등이 있습니다.
      .required(ERROR_MESSAGE.STORE_REQUIRED) // 하단 yup 정리 내용 참고
      .matches(/^[0-9]+$/, ERROR_MESSAGE.STORE_NUMBER)
      .length(10, ERROR_MESSAGE.STORE_LENGTH),
  });

  const submitForm = (values: initialValuesType) => { // 폼을 제출할 때 실행되는 헨들러 내용을 함수로 만들어주세요. 만든 함수는 onSubmit에 넣어주세요.
    console.log(values)
  };

return (<Formik // Formik으로 폼 감싸주기
        initialValues={initialValues} // 폼으로 입력받은 값들 초기값 설정
        validationSchema={schema} // 각 변수마다 validation할 조건들 설정
        onSubmit={submitForm} // 폼 제출시 실행할 헨들러 설정
      >
        {formik => { // formik을 props로 전달
          return (
            <Form> // 자유롭게 Form 작성
              <label className="text-xs font-semibold" htmlFor="storeNumber">
                사업자 등록번호
              </label>
              <Field // 이때 input 대신 Field 사용, input과 동일하게 사용하며, onChange등은 줄 필요가 없고 필수로 type과 name은 지정해주셔야 합니다.
                type="text"
                name="storeNumber" // 이때 name이 곧 각 입력값을 구분하는 변수명이자 아이디가 되므로 초기값 설정한 name과 동일하게 지정해주세요.
                className={`h-12 w-full rounded ${
                  formik.errors.storeNumber ? 'border-red' : 'border-yellow'
                } bg-white pl-3 outline-none focus:border-2`}
                placeholder="1007999997"
              />
              <ErrorMessage // formik에서 제공하는 ErrorMessage를 사용하면 자동으로 validation을 확인해 error일 경우 표시됩니다. (error && <div>에러 메시지</div> 할 필요 x)
                name="storeNumber" // 이때 또한 해당 변수명을 name으로 일치시켜야 합니다.
                component="span" // 에러 메시지는 어떤 형태로 보여줄 지 설정할 수 있습니다. p태그, span태그, div태그 등 다양하게 가능합니다.
                className="w-full text-left text-xs text-red"
              />
              <PrimaryButton type="submit">다음</PrimaryButton> // 버튼 onClick 설정없이 type을 submit으로 설정해주세요.
            </Form>
          );
        }}
      </Formik> )
```
이 외에 좀 더 자세한 내용이나 다른 예제들을 참고하고 싶다면 아래 참고 자료에서 formik 공식 문서와 formik 예제를 참고해주세요.

# yup 라이브러리 소개
Yup은 JavaScript에서 사용되는 객체 스키마 유효성 검사 라이브러리입니다. 이를 통해 간편하게 데이터의 유효성을 확인하고 검증할 수 있습니다.

# yup 사용 방법 예제 정리
```ts
// 에러 메시지 자리에 이 validation을 어겼을 때 띄울 에러 메시지를 넣어주기

//string()
.required(에러 메시지) // 필수 입력란
.matches(보통 정규표현식, 에러 메시지) // 입력 양식 지정
.length(길이, 에러 메시지), // string일 경우 문자열 길이 지정
.min(조건, 에러 메시지) // 최소 문자열 길이 지정
.max(조건, 에러 메시지) // 최대 문자열 길이 지정

//number()
.required(에러 메시지) // 필수 입력란
.positive(에러 메시지) // 보통 변수 타입이 number()라면 양수인지 확인
.negative(에러 메시지) // 음수인지 확인
.integer(에러 메시지) // 정수인지 확인

//array()
.required(에러 메시지) // 필수 입력란
.length(길이, 에러 메시지), // 배열 길이 지정
.min(조건, 에러 메시지) // 최소 배열 길이 지정
.max(조건, 에러 메시지) // 최대 배열 길이 지정
```
그 외의 조건들은 하단의 yub schema 예제 참고 블로그와 yup  제공 기능 종류를 참고해주세요.

# 참고자료
[npm trend 다운로드 그래프](https://npmtrends.com/final-form-vs-formik-vs-react-hook-form-vs-redux-form)
[react hook form vs formik](https://refine.dev/blog/react-hook-form-vs-formik/#formik)

[formik 공식 문서](https://formik.org/docs/overview)
[formik 예제](https://leanylabs.com/blog/form-validation-in-react/#form-validation-library)
[formik 사용방법 설명 블로그](https://abangpa1ace.tistory.com/entry/Form-%EC%A0%9C%EC%9E%91-formik)

[yup npm 공식 문서](https://www.npmjs.com/package/yup)
[yup schema 예제 참고 블로그](https://velog.io/@boyeon_jeong/Yup-%EB%9D%BC%EC%9D%B4%EB%B8%8C%EB%9F%AC%EB%A6%AC-%ED%8C%8C%ED%97%A4%EC%B9%98%EA%B8%B0)
[yup 제공 기능 종류](https://github.com/jquense/yup#arrayminlimit-number--ref-message-string--function-this)

**Comment by jgjgill** [comment]
작성하신 글 잘 읽었습니다..!
읽으면서 궁금한 점이 몇 개 생겨서 글 남깁니다!

- `formik`이 템플릿 코드가 상대적으로 많은데 어느 부분에서 코드를 간결하고 직관적으로 짤 수 있는지가 궁금하긴 하네요.
- 아마도 `react hook form`에서도 context 처럼 전역으로 데이터를 관리할 수 있는 걸로 알고 있습니다..! [useFormContext
](https://react-hook-form.com/docs/useformcontext)

**Reply by Lim-JiSeon** [reply]
- 첫 번째로 코드 길이 부분 수정하겠습니다. 제가 말을 잘못한 것 같습니다. 정확하게 표현하면 다음과 같습니다.
코드 글자 수는 react hook form이 더 적고, 코드 줄 수는 formik이 더 적습니다.(같은 로직과 폼을 구현했을 때)
크게 유의미할 정도의 차이는 아닙니다.
간결하게는 formik에서 제공하는 component를 사용할 경우 더 간결하고 깔끔하게 코드를 짤 수 있습니다.(이 부분은 react hook form도 동일할 것으로 예측)
직관적이란 표현은 react 유저라면 state에 익숙하며 state 기반으로 데이터 흐름이 관리되는 formik이 더 직관적이라 생각했습니다. 다만 이 부분은 사람만다 생각의 관점이 다를 수 있겠다는 생각이 들었습니다. 
-> 이 부분은 제가 잘못 작성한 것 같아 혼란을 드려 죄송합니다. 

- 두 번째로 react hook form 에서도 알려주신 자료로 찾아보니 가능한 것으로 발견했습니다. 이 부분은 이렇게 이해하시면 좋을 것 같습니다. 
formik 같은 경우 <Formik> 으로 감싼 순간 그 안에 데이터 흐름을 context 로 사용 가능합니다. 이유는 formik 자체가 state와 context 기반으로 react 친화적이게 개발되어 나온 이유로 추측됩니다. react hook form 같은 경우는 context로 사용이 가능하다 에 가깝습니다. 
우선 react hook form 이 2가지 조건이 필요하며 그 조건이란 <FormProvider>란 컴포넌트로 감싸줘야 한다는 점 1가지, useFormContext 란 훅을 사용해야 한다는 점 1가지 입니다.

**Reply by Lim-JiSeon** [reply]
저희 그냥 react hook form 사용할까요? ㅎㅎㅎ 
보면 볼수록 2023년도에 react hook form이 너무 많이 달라졌네요.
2023년도부터 react hook form이 앞서 나간 이유가 보니깐 원래 갖고 있던 강점을 살리고, 거기에 formik만의 장점도 가져온 거 같은 느낌이라 뭔가 음... 반박?이라고 해야하나 react hook form이 더 좋은거 같은데 왜 formik을 써야 하나 란 질문에 답을 못 드릴 거 같아서... 보니깐 현재 저만 폼 구현한 거 같은데 저만 바꾸면 될 거 같긴 하네요.

**Reply by Lim-JiSeon** [reply]
아직 종길님 회원가입도 안 하신거 같고 변경한다고 하면 빠르게 변경해야 저도 이번주 내로 수정 가능할 거 같아서... 이 부분은 제가 슬랙에도 언급해놓겠습니당!

**Reply by jgjgill** [reply]
이제 왜 `formik`을 쓰고 `react hook form`과 어떤 차이가 있는지 저희가 이해했으니 선택의 문제인 것 같습니다..!

우선 제가 `react hook form`을 사용하고자 했던 이유는 비제어 컴포넌트 방식의 장점이 크다고 생각해서입니다..!

그럼에도 지금 `formik`을 사용하려고 하는 이유는 지선님이 학습 및 코드를 직접 구성하는 과정을 거쳐서 익숙한 점이 저한테는 장점으로 다가왔습니다.
새로운 기술에 대한 학습 비용 없이 바로 적용할 수 있지 않을까 해서요.

그래서 이 둘의 비용을 따져보고 현재 우리 프로젝트에 폼이 엄청 많다고도 느껴지지 않아서 제가 생각한 방향은 **'빠르게 적용 가능한 걸로 하자'** 입니다!
